extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go

```diff
@@ -309,55 +324,115 @@ func (a *genericActuator) deployMachineDeployments(ctx context.Context, cluster
 // available by the machine-controller-manager. It polls the status every 5 seconds.
-func (a *genericActuator) waitUntilWantedMachineDeploymentsAvailable(ctx context.Context, cluster *controller.Cluster, worker *extensionsv1alpha1.Worker, wantedMachineDeployments worker.MachineDeployments) error {
-	return retryutils.UntilTimeout(ctx, 5*time.Second, 5*time.Minute, func(ctx context.Context) (bool, error) {
-		var numHealthyDeployments, numUpdated, numAvailable, numDesired, numberOfAwakeMachines int32
+func (a *genericActuator) waitUntilWantedMachineDeploymentsAvailable(ctx context.Context, cluster *controller.Cluster, worker *extensionsv1alpha1.Worker, alreadyExistingMachineDeploymentNames sets.String, alreadyExistingMachineClassNames sets.String, wantedMachineDeployments extensionsworker.MachineDeployments, clusterAutoscalerUsed bool) error {
+	rollingUpdate := false
+	autoscalerIsScaledDown := false
+	workerStatusUpdatedForRollingUpdate := false
 
-		// Get the list of all existing machine deployments
-		existingMachineDeployments := &machinev1alpha1.MachineDeploymentList{}
-		if err := a.client.List(ctx, existingMachineDeployments, client.InNamespace(worker.Namespace)); err != nil {
+	return retryutils.UntilTimeout(ctx, 5*time.Second, 3*time.Minute, func(ctx context.Context) (bool, error) {
+		var numHealthyDeployments, numUpdated, numAvailable, numUnavailable, numDesired, numberOfAwakeMachines int32
+
+		// Get the list of all machine deployments
+		machineDeployments := &machinev1alpha1.MachineDeploymentList{}
+		if err := a.client.List(ctx, machineDeployments, client.InNamespace(worker.Namespace)); err != nil {
+			return retryutils.SevereError(err)
+		}
+
+		// Get the list of all machine sets
+		machineSets := &machinev1alpha1.MachineSetList{}
+		if err := a.client.List(ctx, machineSets, client.InNamespace(worker.Namespace)); err != nil {
 			return retryutils.SevereError(err)
 		}
 
+		// map the owner reference to the machine sets
+		ownerReferenceToMachineSet := workerhelper.BuildOwnerToMachineSetMap(machineSets.Items)
+
 		// Collect the numbers of available and desired replicas.
-		for _, existingMachineDeployment := range existingMachineDeployments.Items {
+		for _, deployment := range machineDeployments.Items {
+			wantedDeployment := wantedMachineDeployments.GetDeploymentForName(deployment.Name)
+
 			// Filter out all machine deployments that are not desired (any more).
-			if !wantedMachineDeployments.HasDeployment(existingMachineDeployment.Name) {
+			if wantedDeployment == nil {
 				continue
 			}
 
+			machineSets := ownerReferenceToMachineSet[deployment.Name]
+
+			// use `wanted deployment` for these checks, as the existing deployments can be based on an outdated cache
+			alreadyExistingMachineDeployment := alreadyExistingMachineDeploymentNames.Has(wantedDeployment.Name)
+			newMachineClass := !alreadyExistingMachineClassNames.Has(wantedDeployment.ClassName)
```

Comment:
Why not checking the number of corresponding machine sets? Yes, it also works with checking for classes, but this might break in the future in case we add a "trigger-rolling-update-on-demand" functionality to MCM (no immediate plans, just thinking about it as it might be helpful in some cases). If there is no strong reason against it I'd prefer to use the no. of machine sets indicator.

Reply:
> "trigger-rolling-update-on-demand" 

can you elaborate on this one? I am not sure what you exactly mean.

Reasons for using the MachineClass
- ability to detect a rolling update vs a create operation (you cannot infer that from the machine set alone?)
- the cache of the client that retrieves the machine sets is many times outdated. Alternative would be to use a direct client, but I would prefer not to.

Reply:
> can you elaborate on this one? I am not sure what you exactly mean.

In the future, we might introduce an operation annotation (or similar mechanism) in the MCM/Gardener to allow end-users triggering a rolling update of their worker machines without generating a new machine class.

> ability to detect a rolling update vs a create operation (you cannot infer that from the machine set alone?)

Hm, is this distinction important? What is it used for?

> the cache of the client that retrieves the machine sets is many times outdated. Alternative would be to use a direct client, but I would prefer not to.

I don't get it. Why isn't the cache of the machine classes outdated either?

Reply:
> Hm, is this distinction important? What is it used for? 

It is used so that the RollingUpdate condition is only true if there is a rolling update, not a creation of a pool.

> I don't get it. Why isn't the cache of the machine classes outdated either?

Because we are using the [already existing machine class names](https://github.com/gardener/gardener/blob/c5432f84de767d46d624d5126f45a63df02b3e31/extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go#L94-L93) that we determined before creating the machinedeployments.

> In the future, we might introduce an operation annotation (or similar mechanism) in the MCM/Gardener to allow end-users triggering a rolling update of their worker machines without generating a new machine class.

Mhm, then I guess we need a direct client to avoid cache issues.

Reply:
> It is used so that the RollingUpdate condition is only true if there is a rolling update, not a creation of a pool.

OK, but can't we consider a creation also like a rolling update for the sake of simplicity? I think the boundary conditions are the same (CA should be disabled, etc.), so this way we don't need special code.

> Mhm, then I guess we need a direct client to avoid cache issues.

Well, I'm okay with keeping it as is for now until @tim-ebert has also enhanced the extensions library with the new clientmap concepts. Just saying that checking the machine sets would be more robust.

Reply:
> OK, but can't we consider a creation also like a rolling update for the sake of simplicity? I think the boundary conditions are the same (CA should be disabled, etc.), so this way we don't need special code 

Maybe then we would be better off labeling the condition something like  "UpdateOperation" not "RollingUpdate"?
So any time either a create or rolling update is happening we set the condition to true.

And if in the future we would need to make a distinction between which deployments are actually in creation and which are in rolling update, we could still introducing a condition in the `worker.machineDeployments`.

Reply:
My personal preference is to keep the "RollingUpdate" name, I think it's precise and easy to understand. Yes, we don't remove old nodes in creation because there are none, but IMO it doesn't matter.

Reply:
Why does it make sense to label something "RollingUpdate" if in fact it indicates RollingUpdates + Creations. 
Also what purpose does that serve? 
I feel like this is exactly not precise and confuses external components. 

>  Yes, we don't remove old nodes in creation because there are none

There are also X amount of machine created in parallel unlike in a rolling update.

Reply:
"RollingUpdate" is a well established term and indicates that the nodes are being updated/reconciled. In this case it doesn't matter if it's a creation or a replacement of existing nodes. A more precise name would be something like "NodesReconciliation", but I prefer "RollingUpdate" because it's well-known and used in the community. 
Components that are evaluating this information anyways don't care about the characters inside this string. They just want to know that something is going on with the nodes, e.g. to down-scale the CA. As you explained, there is no use-case where a distinction between create/replacement is required.

Reply:
Alright, I ll call it "RollingUpdate", I guess I am unaware of such a definition. Thanks for the review.

Reply:
I will switch to MachineSets, once Tim Eberts has also enhanced the extensions library with the new clientmap concepts. Shall I open another PR for this or shall this PR stay open in the meanwhile?
cc @tim-ebert

Reply:
I'm not sure, when the ClientMap will make it into the extensions library. So probably you wouldn't want to wait for it, I guess.